File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.isValid(root, -(2 ** 64), (2 ** 64) - 1)

    def isValid(self, root: Optional[TreeNode], min_val: int, max_val: int) -> bool:
        if not root:
            return True
        if root.val >= max_val or root.val <= min_val:
            return False
        return self.isValid(root.left, min_val, root.val) and\
               self.isValid(root.right, root.val, max_val)
        

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root, long min, long max) {
        if (root == nullptr) {
            return true;
        }
        if (root->val >= max || root->val <= min) {
            return false;
        }
        return isValidBST(root->left, min, root->val) 
        && isValidBST(root->right, root->val, max);
    }
    bool isValidBST(TreeNode* root) {
        return isValidBST(root, INT64_MIN, INT64_MAX);
    }
};